cedar-policy-validator/src/schema_file_format.rs

```diff
@@ -133,6 +143,193 @@ impl SchemaFragment {
+        let mut set: HashSet<&Name> = HashSet::from_iter(self.graph.keys().clone());
+        for name in res.iter() {
+            set.remove(name);
+        }
+
+        if let Some(cycle) = set.into_iter().next() {
+            Err(cycle.clone())
+        } else {
+            // We need to reverse the result because, e.g.,
+            // `res` is now [A,B,C] for A -> B -> C because no one depends on A
+            res.reverse();
+            Ok(res)
+        }
+    }
+
+    // Substitute common type references in `ty` according to `resolve_table`
+    fn resolve_type(
+        resolve_table: &HashMap<&Name, SchemaType>,
+        ty: SchemaType,
+        ns: &Option<Name>,
+    ) -> Result<SchemaType> {
+        match ty {
+            SchemaType::TypeDef { type_name } => {
+                let name = if type_name.is_unqualified() {
+                    type_name.prefix_namespace_if_unqualified(ns.clone())
+                } else {
+                    type_name.clone()
+                };
+                resolve_table
+                    .get(&name)
+                    .ok_or(SchemaError::UndeclaredCommonTypes(HashSet::from_iter(
+                        std::iter::once(name.to_string()),
+                    )))
+                    .cloned()
+            }
+            SchemaType::Type(SchemaTypeVariant::Set { element }) => {
+                Ok(SchemaType::Type(SchemaTypeVariant::Set {
+                    element: Box::new(Self::resolve_type(resolve_table, *element, ns)?),
+                }))
+            }
+            SchemaType::Type(SchemaTypeVariant::Record {
+                attributes,
+                additional_attributes,
+            }) => Ok(SchemaType::Type(SchemaTypeVariant::Record {
+                attributes: BTreeMap::from_iter(
+                    attributes
+                        .into_iter()
+                        .map(|(attr, attr_ty)| {
+                            Ok((
+                                attr,
+                                TypeOfAttribute {
+                                    required: attr_ty.required,
+                                    ty: Self::resolve_type(resolve_table, attr_ty.ty, ns)?,
+                                },
+                            ))
+                        })
+                        .collect::<Result<Vec<(_, _)>>>()?,
+                ),
+                additional_attributes,
+            })),
+            _ => Ok(ty),
+        }
+    }
+
+    // Resolve common type references in `self.schema`
+    pub(crate) fn resolve(&mut self) -> Result<()> {
+        let sorted_names = self
+            .topo_sort()
+            .map_err(|name| SchemaError::CycleInCommonTypeReferences(name))?;
+
+        let mut resolve_table = HashMap::new();
+
+        for name in sorted_names.iter() {
+            let ns: Option<Name> = if name.is_unqualified() {
+                None
+            } else {
+                Some(name.namespace().parse().unwrap())
```

Comment:
I think clippy will yell at you for this `unwrap()` without a `PANIC SAFETY` comment

Reply:
Done.